packages/relay/src/lib/eth.ts

```
@@ -1106,57 +1107,66 @@ export class EthImpl implements Eth {
 
+    const parsedTx = await this.parseRawTxAndPrecheck(transaction, requestId);
     const transactionBuffer = Buffer.from(EthImpl.prune0x(transaction), 'hex');
+    const computedHash = EthImpl.prepend0x(createHash('keccak256').update(transactionBuffer).digest('hex'));
+
+    let txSubmitted = false;
     try {
       const contractExecuteResponse = await this.hapiService.getSDKClient().submitEthereumTransaction(transactionBuffer, EthImpl.ethSendRawTransaction, requestId);
-
-      try {
-        // Wait for the record from the execution.
-        const record = await this.hapiService.getSDKClient().executeGetTransactionRecord(contractExecuteResponse, EthereumTransaction.name, EthImpl.ethSendRawTransaction, interactingEntity, requestId);
-        if (!record) {
-          this.logger.warn(`${requestIdPrefix} No record retrieved`);
-          throw predefined.INTERNAL_ERROR();
-        }
-
-        if (record.ethereumHash == null) {
-          this.logger.error(`${requestIdPrefix} The ethereumHash can never be null for an ethereum transaction, and yet it was!!`);
-          throw predefined.INTERNAL_ERROR();
+      txSubmitted = true;
+      // Wait for the record from the execution.
+      let txId = contractExecuteResponse.transactionId.toString();
+      const formattedId = formatTransactionId(txId);
+      const  record = await this.mirrorNodeClient.repeatedRequest('getContractResult', [formattedId], 3, requestId);
+      if (!record) {
+        this.logger.warn(`${requestIdPrefix} No record retrieved`);
+        const tx = await this.mirrorNodeClient.getTransactionById(txId, 0, requestId);
+        if (tx.transactions?.length) {
+          const result = tx.transactions[0].result;
+          if (result === 'WRONG_NONCE') {
+            const accountInfo = await this.mirrorNodeClient.getAccount(parsedTx.from!, requestId)
+            throw predefined.NONCE_TOO_LOW(parsedTx.nonce, accountInfo.ethereumNonce);
```

In case the NONCE is bigger than expected we receive `WRONG_NONCE`, should we return an appropiate error, instead of NONCE_TOO_LOW?

---

Good point, I added a `NONCE_TOO_HIGH` error